docs/proposals/apiserver_watch.md

```diff
@@ -0,0 +1,132 @@
+
+If we decide to support watches spanning different resources in the future and
+we have an efficient indexing mechanisms, it should be relatively simple to unify
+the cache to be common for all the resources.
+
+Since we want the watch in apiserver to be optional for different resource
+types, this needs to be self-contained and hidden behind a well defined API.
+It seems that this API can be very simple - it should just expose a method
+for initializing the cache and another one for registering (and starting) a
+new watcher with a given filtering function (matcher).
+
+The rest of this section describes the concrete steps that need to be done
+to implement the proposal.
+
+1. Since we are initially targeting to have per-resource-type watch in
+apiserver, this should probably be incorporated into generic etcd registry:
+"pkg/registry/generic/etcd". Thus we should migrate all the resources to
+use it (it seems it's not the case now e.g. for services).
+
+2. We should define an API for the watch in apiserver (this will be purely
+internal api just within apiserver, so it should be possible to change it
+relatively easy in the future). Also, we should provide the initial
+implementation that will be just redirecting all the requests to etcd, similarly
+to what we are doing today. This simple implementation should be used to enable
+easy transition to the new mechanism later (api unification).
+
+3. The target implementation internally will have a single watch to etcd open,
+responsible for watching all the changes of objects of a given resource type.
+However, this watch can potentially expire at any time and reconnecting can
+return "too old resource version". In that case relisting is necessary.
+To avoid LIST requests coming from all the watcher at that time, we will
+introduce an additional etcd event type:
+[EtcdResync](../../pkg/tools/etcd_helper_watch.go#L36)
+
+  Whenever reslisting will be done to refresh the internal watch to etcd,
+  EtcdResync event will be send to all the watchers. It will contain the
+  full list of all the objects the watcher is interested in (appropriately
+  filtered) as the parameter of this watch event.
+  Thus, we need to create the EtcdResync event, extend watch.Interface and
+  its implementations to support it and handle those events appropriately
+  in places like
+  [Reflector](../../pkg/client/cache/reflector.go)
+
+4. Create a FIFO cache with a given capacity. This should be as simple as
+having an array an treating it as a cyclic one. Note that in the cache
+we should store two things: the object watched from etcd and its resource
+version (etcdIndex). Obviously resource versions will be increasing (if
+we treat the array as a cyclic one), but they are necessary for registering
+a new watcher that is interested in all the changes that happened after
+a given etcdIndex (resource version).
+
+  We should also support LIST operation from the beginning and the end of
```

Comment:
On Jul 3, 2015, at 2:27 AM, Wojciech Tyczynski <notifications@github.com>
wrote:

In docs/proposals/apiserver_watch.md
<https://github.com/GoogleCloudPlatform/kubernetes/pull/10679#discussion_r33842714>
:

> +  full list of all the objects the watcher is interested in (appropriately
> +  filtered) as the parameter of this watch event.
> +  Thus, we need to create the EtcdResync event, extend watch.Interface and
> +  its implementations to support it and handle those events appropriately
> +  in places like
> +  [Reflector](../../pkg/client/cache/reflector.go)
> +
> +4. Create a FIFO cache with a given capacity. This should be as simple as
> +having an array an treating it as a cyclic one. Note that in the cache
> +we should store two things: the object watched from etcd and its resource
> +version (etcdIndex). Obviously resource versions will be increasing (if
> +we treat the array as a cyclic one), but they are necessary for registering
> +a new watcher that is interested in all the changes that happened after
> +a given etcdIndex (resource version).
> +
> +  We should also support LIST operation from the beginning and the end of

I probably didn't understand your suggestion - but what if watch window
expires both in our cache and in etcd - the we simply cannot rewatch from
that particular time and we need some kind of relisting?


------

We can also support "last seen" for a watch. Imagine a watch opened on
"services" and one opened on "services/test" (by name).  Once we start
watching "services", we know "test" implicitly does or does not change by
observing it is not updated.  Therefore, if we track the last resource
version seen in a map for that particular watch (t1), if the
"services/test" watch is disconnected, then reconnects with an older
resource version (t1) than that last version seen, we know we can start the
watch for the client at the latest resource version we have seen for *any*
service.

This is another reason for having all the watches on a single watch on etcd
across all resources.  Our most common LIST comes from watches to
infrequently changed resource types in the presence of rapid updates in
other types - example, services rarely change but events are firing all the
time, so the odds of a services client being able to reconnect within the
window are low. However, if we observe most resources with one watch
(potentially excluding events) then we can easily track the HWM resource
version for each type or watch, and given that we know when we can tell
clients to watch from latest, vs relisting.

------

After some thinking, I agree that we probably don't need the list from the
"beginning time". But the list from "end time" of the cache is supposed to
handle exactly this case. We can do relisting by sending "Resync" event to
client, but for this we need to send him all objects (filtered) and to do
it effectively we need to now the list of all objects - this is why I'm
mentioning this LIST stuff.


-------

While I originally thought that sending the list of events through watch
would make clients easier, I think it has some limitations.  First, all
clients have to support a new type of event, even those that don't want
it.  Second, we may eventually want to paginate that list for a client (50K
pods at 1K each in one watch event?), which means sending multiple resync
events in sequence with starting and closing.  I'd really prefer that to be
part of a list API (since it's also useful there).

If we want to serve list results from memory, we can do that this way, or
by operating a higher level cache.  Or we could just run an etcd proxy (in
process), which would do all that for us and be able to serve stale reads
as well.  The TCP churn for rewatches is significant - but if we do plan to
mirror etcd it might be worth asking whether we just want a local etcd
proxy (the one that maintains the state map, not an HTTP proxy).

-------

However, I didn't understand your "high water" suggestion - can you please
explain it a bit more?

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/pull/10679/files#r33842714>
.